### ISSUE TITLE
NET-1565 Fix S1192 FP: Shouldn't raise on Dapper parameters

### ISSUE THREAD

**Issue by mgirolet-gl** [SonarSource/sonar-dotnet#9569]
### Description

Issue S1192 "String literals should not be duplicated" raises when a string literal is duplicated more than a given amount of time (3 by default), excluding literals with <5 characters, ones matching a parameter name and ones used in attributes.

I think it also shouldn't raise in the case of Dapper parameters used for SQL Requests.

The reasoning behind this is that the Dapper parameters are given a name that needs to be the same as the one in the following SQL request to be executed with these parameters.

However, the "Why is this an issue?" claims: "[...], as any change would need to be propagated on all occurrences.", which is not the case here: renaming a Dapper parameter in a function only means it has to be renamed in the following SQL request, NOT in all occurences of that string literal. Therefore, it doesn't affect maintenability in the way described by the issue.

Therefore, I think an exclusion should be created for this case. Dapper parameters start with an '@', maybe can this be used as a criteria?

### Repro steps

In the below example with monsters, players and dungeon all identified by a name, if we one day decide to change players to make them be identified by an ID, it doesn't require changing all occurences of the ``"@name"`` string, but only the ones in the ``GetPlayer`` function.

Which defeats the point of S1192 claiming that this literal would need to have all its occurences manually (therefore painfully) updated if one of them gets changed, which is not (and most of the time never) the case with Dapper parameters.

This example isn't a niche case: there are lots of parameter names for SQL requests that are very common but all refer to different things.
Even in the case of a parameter refering to the same column of the same table in two different functions, changing it in one function for any reason does not mean that other occurences of that literal (even in the same context) would need to be changed.

```cs
using Dapper;

// Assuming db is a System.Data.Common.DbConnection connected to a MySQL database

// Assuming Monster, Player and Dungeon are objects

public Monster? GetMonster(string name)
{
	string req = $@"
		SELECT 
			*
		FROM 
			monsters
		WHERE 
			name = @name
		;";

	DynamicParameters param = new DynamicParameters();
	param.Add("@name", name); // Raises because "@name" has been duplicated too many times, even though changing this literal would only require changing the previously written request and not all occurences, therefore having no impact on maintenability
	return db.Query<Monster>(req, param).FirstOrDefault();
}

public Player? GetPlayer(string name)
{
	string req = $@"
		SELECT 
			*
		FROM 
			players
		WHERE 
			name = @name
		;";

	DynamicParameters param = new DynamicParameters();
	param.Add("@name", name); // Raises because "@name" has been duplicated too many times, even though changing this literal would only require changing the previously written request and not all occurences, therefore having no impact on maintenability
	return db.Query<Player>(req, param).FirstOrDefault();
}

public Dungeon? Function3(string name)
{
	string req = $@"
		SELECT 
			*
		FROM 
			dungeon
		WHERE 
			name = @name
		;";

	DynamicParameters param = new DynamicParameters();
	param.Add("@name", name); // Raises because "@name" has been duplicated too many times, even though changing this literal would only require changing the previously written request and not all occurences, therefore having no impact on maintenability
	return db.Query<Dungeon>(req, param).FirstOrDefault();
}
```

### Expected behavior

I'd expect the rule not to raise for Dapper parameter names

### Actual behavior

The issue raises because it assumes it is a literal that has the same meaning everywhere it is repeated

### Known workarounds

Besides making a constant that will make your requests harder to maintain, or giving different parameter names to each and every request that has parameters with a similar name, there isn't really a workaround.

### Related information

* C#/VB.NET Plugins version: 9.27.0.93347
* Visual Studio version: 2022
* MSBuild version: 17.7.6
* Dotnet version: .NET 6.0
* SonarScanner for .NET version: 6.2.0
* Operating System: Windows/Linux

**Comment by zsolt-kolbay-sonarsource** [comment]
Hi @mgirolet-gl. Thank you for the detailed explanation and the code samples. I confirmed this as a False Positive and added it to our backlog.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1565](https://sonarsource.atlassian.net/browse/NET-1565)

[NET-1565]: https://sonarsource.atlassian.net/browse/NET-1565?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by alex-meseldzija-sonarsource** [comment]
Hi @mgirolet-gl ,
this has been fixed in the 10.23 [release](https://github.com/SonarSource/sonar-dotnet-enterprise/releases/tag/10.23.0.137933) of our analyzers.
